verification/src/transaction_verifier.rs

```diff
@@ -20,6 +20,59 @@ use lru::LruCache;
 use std::cell::RefCell;
 use std::collections::HashSet;
 
+/// The phase that transactions are in.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum TransactionVerificationPhase {
+    /// The transaction has just been submitted.
+    ///
+    /// So the transaction will be:
+    /// - proposed after the `tip_number + 1` block.
+    /// - committed after the `tip_number + 1 + 2` block.
+    Submitted,
+    /// The transaction has already been proposed and in Gap.
+    ///
+    /// So the transaction will be committed after the `tip_number + 2` block.
+    Gap,
+    /// The transaction has already been proposed.
+    ///
+    /// So the transaction will be committed after the `tip_number + 1` block.
+    Proposed,
+    /// The transaction is commit.
+    ///
+    /// So the transaction will be committed in current block.
+    Committed,
+}
+
+impl TransactionVerificationPhase {
+    /// Convert tip block number to earliest committed block number.
+    pub fn earliest_committed_block(self, tip_block_number: BlockNumber) -> BlockNumber {
+        match self {
+            Self::Submitted => tip_block_number + 3,
+            Self::Gap => tip_block_number + 2,
+            Self::Proposed => tip_block_number + 1,
+            Self::Committed => tip_block_number,
+        }
+    }
+
+    /// Convert tip epoch number to earliest committed epoch number.
+    pub fn earliest_committed_epoch(self, tip_epoch: EpochNumberWithFraction) -> EpochNumber {
+        let number = tip_epoch.number();
+        let length = tip_epoch.length();
+        let index = tip_epoch.index();
+        let index_added = match self {
+            Self::Submitted => 3,
+            Self::Gap => 2,
```

Comment:
The gap is not a constant but is configurable in the consensus via the option `tx_proposal_window`